src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,180 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Playwright.Transport
+{
+    internal class WebSocketTransport : IConnectionTransport, IDisposable
+    {
+        private const int DefaultBufferSize = 16000;  // Byte buffer size
+        private readonly ClientWebSocket _webSocket;
+        private readonly string _wsEndpoint;
+        private readonly CancellationTokenSource _readerCancellationSource;
+        private readonly BrowserTypeConnectOptions _options;
+        private readonly int _slowMo;
+        private readonly int _timeout;
+
+        internal WebSocketTransport(
+            string wsEndpoint = default,
+            BrowserTypeConnectOptions options = default)
+        {
+            _webSocket = new ClientWebSocket();
+            _wsEndpoint = wsEndpoint;
+            _options = options;
+            _slowMo = _options?.SlowMo != null ? (int)_options?.SlowMo : 0;
+            _timeout = _options?.Timeout != null ? (int)_options?.Timeout : 30000;
+            _readerCancellationSource = new CancellationTokenSource(_timeout);
+            SetRequestHeaders();
+            _ = ConnectAsync();
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        ~WebSocketTransport() => Dispose(false);
+
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        public event EventHandler<LogReceivedEventArgs> LogReceived;
+
+        public event EventHandler<TransportClosedEventArgs> TransportClosed;
+
+        public bool IsClosed { get; private set; }
+
+        public async Task SendAsync(string message)
+        {
+            await Task.Delay(_slowMo).ConfigureAwait(false);
+
+            try
+            {
+                var messageBuffer = Encoding.UTF8.GetBytes(message);
+                if (!_readerCancellationSource.IsCancellationRequested)
+                {
+                    await _webSocket.SendAsync(new ArraySegment<byte>(messageBuffer, 0, messageBuffer.Length), WebSocketMessageType.Text, true, _readerCancellationSource.Token).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Close(string closeReason)
+        {
+            if (!IsClosed)
+            {
+                IsClosed = true;
+                HandleSocketClose(closeReason);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Close(Exception ex)
+        {
+            System.Diagnostics.Debug.WriteLine(ex);
+            Close(ex.ToString());
+        }
+
+        private void HandleSocketClose(string closeReason)
+        {
+            TransportClosed?.Invoke(this, new TransportClosedEventArgs { CloseReason = closeReason });
+            _readerCancellationSource?.Cancel();
```

Comment:
We are handling the socket close, so all the socket operations will fail, I don't think you need this Cancel() anymore?

Reply:
yup not required

Reply:
It is required, if there are any task queued that needs to be cancelled since socket is closed. It is good to have. We do similar thing for `StdIOTransport` - https://github.com/microsoft/playwright-dotnet/blob/41740926d374220d1becd95756cc62099b425d52/src/Playwright/Transport/StdIOTransport.cs#L75

Reply:
Anyways, this reader token seems to be redundant. Removing all together.

Reply:
updated